include/fiction/algorithms/physical_design/graph_oriented_layout_design.hpp

```diff
@@ -0,0 +1,1574 @@
+//
+// Created by Simon Hofmann on 30.01.24.
+//
+#ifndef FICTION_GRAPH_ORIENTED_LAYOUT_DESIGN_HPP
+#define FICTION_GRAPH_ORIENTED_LAYOUT_DESIGN_HPP
+
+#include "fiction/algorithms/network_transformation/fanout_substitution.hpp"
+#include "fiction/algorithms/path_finding/a_star.hpp"
+#include "fiction/algorithms/path_finding/cost.hpp"
+#include "fiction/algorithms/path_finding/distance.hpp"
+#include "fiction/algorithms/physical_design/post_layout_optimization.hpp"
+#include "fiction/io/print_layout.hpp"
+#include "fiction/layouts/bounding_box.hpp"
+#include "fiction/layouts/clocking_scheme.hpp"
+#include "fiction/layouts/obstruction_layout.hpp"
+#include "fiction/networks/technology_network.hpp"
+#include "fiction/traits.hpp"
+#include "fiction/types.hpp"
+#include "fiction/utils/name_utils.hpp"
+#include "fiction/utils/network_utils.hpp"
+#include "fiction/utils/placement_utils.hpp"
+#include "fiction/utils/routing_utils.hpp"
+
+#include <fmt/format.h>
+#include <mockturtle/traits.hpp>
+#include <mockturtle/utils/node_map.hpp>
+#include <mockturtle/utils/stopwatch.hpp>
+#include <mockturtle/views/fanout_view.hpp>
+#include <mockturtle/views/immutable_view.hpp>
+#include <mockturtle/views/names_view.hpp>
+
+#include <array>
+#include <cassert>
+#include <chrono>
+#include <cstddef>
+#include <cstdint>
+#include <functional>
+#include <iostream>
+#include <optional>
+#include <queue>
+#include <string>
+#include <tuple>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace fiction
+{
+
+/**
+ * Parameters for the graph-oriented layout design algorithm.
+ */
+struct graph_oriented_layout_design_params
+{
+    /**
+     * In high effort mode, 12 search space graphs are created with varying fanout substitution strategies, allowed PI
+     * placements, and other parameters, compared to only 2 graphs in high efficiency mode. This broader exploration
+     * increases the likelihood of discovering optimal layouts. When a solution is found in any graph, its cost is used
+     * to prune the search in the remaining graphs.
+     */
+    bool high_effort_mode = false;
+    /**
+     * Return the first found layout, which might still have a high cost but can be found fast.
+     */
+    bool return_first = false;
+    /**
+     * Timeout limit (in ms).
+     */
+    uint64_t timeout = std::numeric_limits<uint64_t>::max();
```

Comment:
When running in high-efficiency mode with no timeout set, GOLD might take forever to complete, even on rather small functions. Maybe it is useful to define a maximum number of paths to evaluate in high-efficiency mode. Without having put much thought into it, maybe we can say explore 1% of paths and then return the best result. While the high-effort mode would evaluate all.

Reply:
Just an inspiration on how to handle this situation. Your solution may be different of course.

Reply:
I have another proposal:

In high-effiency mode, the timeout is set to 10s.
After the 10s, the best found layout is returned, but if no solution was found, the timeout is ignored until the first solution is found.
This behavior is of course disable when the user sets a timeout.

Reply:
Sounds good. Let's try this! :muscle: